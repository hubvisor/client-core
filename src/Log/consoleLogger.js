import { hashCode } from '../util'

// generates a unique color for each tag according to the string to quickly identify each sub-system tag
function colorFromTag (tag) {
  if (!tag) { return '#607D8B' }

  const angle = Math.PI // because why not
  // hue depend on tag hash, saturation and value are hand-picked from a flat color
  const hue = (hashCode(tag) * angle) % 360
  const saturation = 70
  const value = 53
  return `hsl(${hue}, ${saturation}%, ${value}%)`
}

const levelColors = {
  debug: '#9C27B0',
  info: '#4CAF50',
  warn: '#FF9800',
  error: '#F44336'
}

function createBadge ({ level, name, tag }) {
  const tagColor = colorFromTag(tag)
  const levelColor = levelColors[level] || '#000'
  const badge = [
    {
      value: name,
      style: 'background: linear-gradient(#4D4D4D, #5D5D5D); color: #fff; padding: 2px 4px; border-top-left-radius: 4px; border-bottom-left-radius: 4px;'
    }, {
      value: level,
      style: `background: ${levelColor}; color: #fff; padding: 2px 4px;`
    }, {
      value: tag,
      style: `background: ${tagColor}; color: #fff; padding: 2px 2px; border-top-right-radius: 4px; border-bottom-right-radius: 4px;`
    }
  ].filter(item => item.value)

  const badgeFormatString = badge.map(item => `%c${item.value}`).join('')
  const badgeArgs = badge.map(item => item.style)
  return [ badgeFormatString, ...badgeArgs ]
}

export default function consoleLogger ({ level, name, tag, items }) {
  const consoleOutputs = {
    table: console.table,
    warn: console.warn,
    error: console.error
  }
  const badge = level === 'table' ? [] : createBadge({ name, level, tag })
  const output = consoleOutputs[level] || console.log

  return output(...badge, ...items)
}
