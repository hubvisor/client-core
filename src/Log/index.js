import { autobind } from '../util'
import consoleLog from './consoleLogger'

// because of https://github.com/jsdoc3/jsdoc/issues/1132
/** @module Log */

/**
 * This callback is displayed as part of the Requester class.
 * @callback Log~Logger
 * @param {Object} message The message to be logged
 * @param {string} message.level The log level to use for this message
 * @param {string} message.name The log name to use for this message
 * @param {string} message.tag The log tag to use for this message
 * @param {any[]} items The elements to log
 */

/**
 * Logs messages to a logger implementation
 */
export default class Log {
  static console ({ name, tag } = {}) { return new Log({ name, tag, logger: consoleLog }) }

  static noop ({ name, tag } = {}) { return new Log({ name, tag }) }

  /**
   * @param {Object} [options] The options for this log
   * @param {string} [options.name] The name of the log
   * @param {string} [options.tag] An optional tag (for example to display a specific subsystem messages)
   * @param {Log~Logger} [options.logger] An optional logger to direct outputs to. If no value is provided, logs functions will be noops
   */
  constructor ({ name, tag, logger } = {}) {
    this._name = name
    this._tag = tag
    this._logger = logger

    const noop = () => {}
    const log = (level, items) => logger({ level, name, tag, items })
    this._log = logger ? log : noop

    autobind(this)
  }

  /**
   * Prints a message with verbose level
   * @param {...*} items The items to log
   */
  trace (...items) { this._log('trace', items) }

  /**
   * Prints a message with debug level
   * @param {...*} items The items to log
   */
  debug (...items) { this._log('debug', items) }

  /**
   * Prints a message with normal level
   * @param {...*} items The items to log
   */
  info (...items) { this._log('info', items) }

  /**
   * Prints a message with warning level
   * @param {...*} items The items to log
   */
  warn (...items) { this._log('warn', items) }

  /**
   * Prints a message with error level
   * @param {...*} items The items to log
   */
  error (...items) { this._log('error', items) }

  /**
   * Displays data in a tabular form, @see {@link https://developer.mozilla.org/fr/docs/Web/API/Console/table}
   * @param {object|array} data The data to see in a table
   */
  table (data) { this._log('table', [ data ]) }

  /**
   * Instantiates a new Logger instance with a given tag
   * @param {string} tag The tag to create the logger with
   * @returns {Logger} The new logger using the provided tag
   */
  tag (tag) {
    const { _name: name, _logger: logger } = this
    return new Log({ name, tag, logger })
  }
}
