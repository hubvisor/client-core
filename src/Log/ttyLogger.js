import { hashCode } from '../util'
import chalk from 'chalk'

// ttyLogger can be used in tests !

// generates a unique color for each tag according to the string to quickly identify each sub-system tag
function colorFromTag (tag) {
  if (!tag) { return [ 200, 31, 55 ] }

  const angle = Math.PI // because why not
  // hue depend on tag hash, saturation and value are hand-picked from a flat color
  const hue = (hashCode(tag) * angle) % 360
  const saturation = 70
  const value = 53

  return [ hue, saturation, value ]
}

const levelColors = {
  debug: '#9C27B0',
  info: '#4CAF50',
  warn: '#FF9800',
  error: '#F44336'
}

function createBadge ({ level, name, tag }) {
  const tagColor = colorFromTag(tag)
  const levelColor = levelColors[level] || '#00000'
  const items = [
    {
      value: ` ${name} `,
      formatter: chalk.bold.hex('#ffffff').bgHex('#4D4D4D')
    }, {
      value: ` ${level.toUpperCase()} `,
      formatter: chalk.bold.hex('#ffffff').bgHex(levelColor)
    }, {
      value: ` ${tag} `,
      formatter: chalk.bold.hex('#ffffff').bgHsl(...tagColor)
    }
  ]
    .filter(item => item.value)
    .map(({ value, formatter }) => formatter(value))

  return [ items.join('') ]
}

export default function consoleLogger ({ level, name, tag, items }) {
  const consoleOutputs = {
    table: console.table
  }
  const badge = level === 'table' ? [] : createBadge({ name, level, tag })
  const output = consoleOutputs[level] || console.log
  return output(...badge, ...items)
}
