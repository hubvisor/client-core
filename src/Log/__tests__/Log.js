import Log from '..'

describe('Log', () => {
  let consoleLog
  let consoleWarn
  let consoleError

  beforeEach (() => {
    consoleLog = jest.spyOn(console, 'log')
    consoleWarn = jest.spyOn(console, 'warn')
    consoleError = jest.spyOn(console, 'error')
    const mocks = [ consoleLog, consoleWarn, consoleError ]

    // hide logs while running these tests
    mocks.forEach(mock => mock.mockImplementation(() => { /* do nothing */ }))
  })

  afterEach (() => {
    [ consoleLog, consoleWarn, consoleError ].forEach(mock => {
      mock.mockReset()
      mock.mockRestore()
    })
  })

  const logArgs = [ 'foo', 'bar', 42, { foo: 'bar' } ]
  const badge = 1
  const nameColor = 1
  const levelColor = 1
  const tagColor = 1

  test('without name', () => {
    const log = Log.console()
    log.info(...logArgs)
    expect(consoleLog).toMatchSnapshot()
  })

  test('with name', () => {
    const log = Log.console({ name: 'TEST' })
    log.info(...logArgs)
    expect(consoleLog).toMatchSnapshot()
  })

  test('no tag', () => {
    const log = Log.console({ name: 'TEST' })
    log.info(...logArgs)
    expect(consoleLog).toMatchSnapshot()
  })

  test('one tag', () => {
    const log = Log.console({ name: 'TEST' })
    log.tag('tag').info(...logArgs)
    expect(consoleLog).toMatchSnapshot()
  })

  test('chain tag', () => {
    const log = Log.console({ name: 'TEST' })
    log.tag('tag1').tag('tag2').info(...logArgs)
    expect(consoleLog).toMatchSnapshot()
  })

  test('levels', () => {
    const log = Log.console()
    log.trace('test')
    log.debug('test')
    log.info('test')

    expect(consoleLog).toMatchSnapshot()
    expect(consoleWarn).toMatchSnapshot()
    expect(consoleError).toMatchSnapshot()
  })

  test('noop log', () => {
    const log = Log.noop()
    log.trace('test')
    log.debug('test')
    log.info('test')

    expect(consoleLog).toMatchSnapshot()
    expect(consoleWarn).toMatchSnapshot()
    expect(consoleError).toMatchSnapshot()
  })
})
