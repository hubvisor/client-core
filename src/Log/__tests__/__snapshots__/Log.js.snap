// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log chain tag 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cTEST%cinfo%ctag2",
      "background: linear-gradient(#4D4D4D, #5D5D5D); color: #fff; padding: 2px 4px; border-top-left-radius: 4px; border-bottom-left-radius: 4px;",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "background: hsl(335.68956412002444, 70%, 53%); color: #fff; padding: 2px 2px; border-top-right-radius: 4px; border-bottom-right-radius: 4px;",
      "foo",
      "bar",
      42,
      Object {
        "foo": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log levels 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%ctrace",
      "background: #000; color: #fff; padding: 2px 4px;",
      "test",
    ],
    Array [
      "%cdebug",
      "background: #9C27B0; color: #fff; padding: 2px 4px;",
      "test",
    ],
    Array [
      "%cinfo",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "test",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log levels 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cwarn",
      "background: #FF9800; color: #fff; padding: 2px 4px;",
      "test",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log levels 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cerror",
      "background: #F44336; color: #fff; padding: 2px 4px;",
      "test",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log no tag 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cTEST%cinfo",
      "background: linear-gradient(#4D4D4D, #5D5D5D); color: #fff; padding: 2px 4px; border-top-left-radius: 4px; border-bottom-left-radius: 4px;",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "foo",
      "bar",
      42,
      Object {
        "foo": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log noop log 1`] = `[MockFunction]`;

exports[`Log noop log 2`] = `[MockFunction]`;

exports[`Log noop log 3`] = `[MockFunction]`;

exports[`Log one tag 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cTEST%cinfo%ctag",
      "background: linear-gradient(#4D4D4D, #5D5D5D); color: #fff; padding: 2px 4px; border-top-left-radius: 4px; border-bottom-left-radius: 4px;",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "background: hsl(342.53580424003303, 70%, 53%); color: #fff; padding: 2px 2px; border-top-right-radius: 4px; border-bottom-right-radius: 4px;",
      "foo",
      "bar",
      42,
      Object {
        "foo": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log with name 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cTEST%cinfo",
      "background: linear-gradient(#4D4D4D, #5D5D5D); color: #fff; padding: 2px 4px; border-top-left-radius: 4px; border-bottom-left-radius: 4px;",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "foo",
      "bar",
      42,
      Object {
        "foo": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Log without name 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "%cinfo",
      "background: #4CAF50; color: #fff; padding: 2px 4px;",
      "foo",
      "bar",
      42,
      Object {
        "foo": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
